File: practice-01/exercise-05/go/main.go
package main

import (
	"fmt"
	"strings"
)

const N = 3

type Article struct {
	code  int
	name  string
	price float32
	stock int
}

func load(x *[N]Article) {
	for i := 0; i < N; i++ {
		fmt.Println("Código: ")
		fmt.Scan(&x[i].code)
		fmt.Println("Nombre: ")
		fmt.Scan(&x[i].name)
		fmt.Println("Precio unitario: ")
		fmt.Scan(&x[i].price)
		fmt.Println("Stock: ")
		fmt.Scan(&x[i].stock)
	}
}

func searchByName(x *[N]Article, name string) int {
	var i = 0
	for i < N && x[i].name != name {
		i++
	}
	return i
}

func searchByCode(x *[N]Article, code int) int {
	var i = 0
	for i < N && x[i].code != code {
		i++
	}
	return i
}

func process(x *[N]Article) {
	var articleName string
	var amount int

	fmt.Print("Nombre del producto: ")
	fmt.Scan(&articleName)

	for strings.ToLower(articleName) != "fin" {
		fmt.Print("Unidades vendidas: ")
		fmt.Scan(&amount)

		idx := searchByName(x, articleName)

		if idx < N {
			x[idx].stock -= amount
		} else {
			fmt.Println("No encontrado.")
		}

		fmt.Print("Nombre del producto: ")
		fmt.Scan(&articleName)
	}
}

func checkEmptyStock(x *[N]Article) {
	fmt.Println("Articulos sin stock: ")
	for _, a := range x {
		if a.stock == 0 {
			fmt.Printf("	- %s\n", a.name)
		}
	}
}

func showStock(x *[N]Article) {
	var code int

	fmt.Print("Ingrese el código: ")
	fmt.Scan(&code)

	idx := searchByCode(x, code)
	if idx < N {
		fmt.Printf("Stock: %d\n", x[idx].stock)
	} else {
		fmt.Println("No encontrado.")
	}
}

func main() {
	var articles = [N]Article{
		{code: 101, name: "Lapicera", price: 150.50, stock: 20},
		{code: 102, name: "Cuaderno", price: 300.00, stock: 10},
		{code: 103, name: "Regla", price: 120.75, stock: 5},
	}
	process(&articles)
	fmt.Println()
	checkEmptyStock(&articles)
	fmt.Println()
	showStock(&articles)
}

File: practice-01/exercise-05/c/main.c
#include <stdio.h>
#include <string.h>
#define M 50
#define N 3

struct Article
{
  int code;
  char name[M];
  float unitary_price;
  int stock;
};

void load(struct Article x[])
{
  for (int i = 0; i < N; i++)
  {
    printf("Codigo: ");
    scanf("%d", &x[i].code);
    printf("Nombre: ");
    fgets(x[i].name, sizeof(char) * M, stdin);
    while (getchar() != '\n')
      ;
    printf("Precio unitario: ");
    scanf("%f", &x[i].unitary_price);
    printf("Stock: ");
    scanf("%d", &x[i].stock);
    printf("\n");
  }
}

int searchByCode(struct Article x[], int code)
{
  int i = 0;
  while (i < N && x[i].code != code)
    i++;
  return i;
}

int searchByName(struct Article x[], char *name)
{
  int i = 0;

  name[strcspn(name, "\n")] = '\0';
  while (i < N && strcasecmp(x[i].name, name) != 0)
    i++;
  return i;
}

void process(struct Article x[])
{
  char articleName[M];
  int amount;

  printf("Nombre del producto: ");
  fgets(articleName, sizeof(char) * M, stdin);
  articleName[strcspn(articleName, "\n")] = '\0';

  while (strcasecmp(articleName, "FIN") != 0)
  {
    printf("Cantidad de unidades vendidas: ");
    scanf("%d", &amount);
    while (getchar() != '\n')
      ;

    int idx = searchByName(x, articleName);

    if (idx < N)
    {
      x[idx].stock -= amount;
    }
    else
    {
      printf("No econtrado.\n");
    }

    printf("Nombre del producto: ");
    fgets(articleName, sizeof(char) * M, stdin);
    articleName[strcspn(articleName, "\n")] = '\0';
  }
}

void checkEmptyStock(struct Article x[])
{
  printf("Articulos sin stock:\n");
  for (int i = 0; i < N; i++)
  {
    if (x[i].stock == 0)
    {
      printf("  - %s\n", x[i].name);
    }
  }
}

void showStock(struct Article x[])
{
  int code;
  printf("Ingrese el codigo: ");
  scanf("%d", &code);

  int idx = searchByCode(x, code);
  if (idx < N)
  {
    printf("Stock: %d", x[idx].stock);
  }
  else
  {
    printf("No econtrado.");
  }
}

void main()
{
  // struct Article articles[N];
  struct Article articles[N] = {
      {101, "Lapicera", 150.50, 20},
      {102, "Cuaderno", 300.00, 10},
      {103, "Regla", 120.75, 0}};

  // load(articles);
  // printf("\n");
  process(articles);
  printf("\n");
  checkEmptyStock(articles);
  printf("\n");
  showStock(articles);
}